modules/500-upmeter/hooks/dynamic_probe/hook.go

```diff
@@ -0,0 +1,153 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package dynamic_probe
+
+import (
+	"github.com/flant/addon-operator/pkg/module_manager/go_hook"
+	"github.com/flant/addon-operator/sdk"
+	"github.com/flant/shell-operator/pkg/kube_events_manager/types"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"sigs.k8s.io/yaml"
+
+	"github.com/deckhouse/deckhouse/go_lib/set"
+)
+
+// This hook populates internal values with object names that are used for dynamic probes in upmeter,
+// and are rendered in its Helm templates.
+var _ = sdk.RegisterFunc(
+	&go_hook.HookConfig{
+		OnBeforeHelm: &go_hook.OrderedConfig{Order: 10},
+		Queue:        "/modules/upmeter/dynamic_probes",
+		Kubernetes: []go_hook.KubernetesConfig{
+			{
+				Name:       "upmeter_discovery_ingress_controllers",
+				ApiVersion: "v1",
+				Kind:       "ConfigMap",
+				NameSelector: &types.NameSelector{
+					MatchNames: []string{"upmeter-discovery-controllers"},
+				},
+				NamespaceSelector: &types.NamespaceSelector{
+					NameSelector: &types.NameSelector{
+						MatchNames: []string{"d8-ingress-nginx"},
+					},
+				},
+				FilterFunc: filterNamesFromConfigmap,
+			},
+			{
+				Name:       "upmeter_discovery_nodegroups",
+				ApiVersion: "v1",
+				Kind:       "ConfigMap",
+				NameSelector: &types.NameSelector{
+					MatchNames: []string{"upmeter-discovery-nodegroups"},
+				},
+				NamespaceSelector: &types.NamespaceSelector{
+					NameSelector: &types.NameSelector{
+						MatchNames: []string{"d8-system"},
+					},
+				},
+				FilterFunc: filterNamesFromConfigmap,
+			},
+			{
+				Name:       "cloud_provider_secret",
+				ApiVersion: "v1",
+				Kind:       "Secret",
+				NameSelector: &types.NameSelector{
+					MatchNames: []string{"d8-node-manager-cloud-provider"},
+				},
+				NamespaceSelector: &types.NamespaceSelector{
+					NameSelector: &types.NameSelector{
+						MatchNames: []string{"kube-system"},
+					},
+				},
+				FilterFunc: filterCloudProviderAvailabilityZonesFromSecret,
+			},
+		},
+	},
+
+	collectDynamicNames,
+)
+
+// collectDynamicNames sets names of objects to internal values
+func collectDynamicNames(input *go_hook.HookInput) error {
+	// Input, empty strings mean invalidated data
+	var (
+		ingressNames   = parseSingleStringSet(input.Snapshots["upmeter_discovery_ingress_controllers"]).Delete("").Slice()
+		nodeGroupNames = parseSingleStringSet(input.Snapshots["upmeter_discovery_nodegroups"]).Delete("").Slice()
+		cloudZones     = parseSingleStringSet(input.Snapshots["cloud_provider_secret"]).Delete("").Slice()
+	)
+
+	// Populate values
+	data := emptyNames().WithIngressControllers(ingressNames...)
```

Comment:
Why don't you check if ingressNames len is 0? Like cloudZones.

Reply:
We have to sets of probes: ingress-controllers and node-groups.

We can iterate over ingress-controllers anyway, even if the list is empty.

But we avoid adding nodegroups probes if we have no cloud zones, because the probe checks minPerZone, and if zones are not set, we don't believe we can check anything. So, zones and nodegroups come together. Empty zone list is a guard for nodegroup probes. 

Otherwise, iterating over empty lists is not a problem, either for controllers or nodegroups.